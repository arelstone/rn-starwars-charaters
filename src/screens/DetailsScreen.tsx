import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import React, { FC, useEffect } from 'react';
import { View, Text } from 'react-native';
import { useSelector } from 'react-redux';
import { AppNavigatorParams } from '../navgators';
import { RootState } from '../store';

interface DetailsScreenProps {
    navigation: StackNavigationProp<AppNavigatorParams, 'Details'>;
    route: RouteProp<AppNavigatorParams, 'Details'>;
}

const DetailsScreen: FC<DetailsScreenProps> = ({ navigation, route }) => {
    const person = useSelector((state: RootState) => state.persons.results.find(person => person.id === route.params.id));

    useEffect(() => {
        navigation.setOptions({ title: person!.name });
    }, [navigation, person]);



    return <View><Text>DetailsScreen</Text></View>;
};



export default DetailsScreen;
