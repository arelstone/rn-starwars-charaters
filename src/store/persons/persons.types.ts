import { Person } from '../../SwApi';

export const START_FETCHING_PERSONS = 'START_FETCHING_PERSONS';
export const DONE_FETCHING_PERSONS = 'DONE_FETCHING_PERSONS';
export const SET_PERSONS = 'SET_PERSONS';
export const FAILED_FETCHING_PERSONS = 'FAILED_FETCHING_PERSONS';

export type SetPersonsParams = {
    count: number;
    nextPage: number;
    results: Person[];
}

export type FetchPersonsAction = {
    type: PersonActionTypes;
}

type PersonActionTypes = typeof START_FETCHING_PERSONS
    | typeof DONE_FETCHING_PERSONS
    | typeof FAILED_FETCHING_PERSONS;

export type SetPersonsAction = {
    type: typeof SET_PERSONS;
    payload: SetPersonsParams;
}
