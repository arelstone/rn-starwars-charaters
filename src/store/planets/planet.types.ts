import { Planet } from '../../SwApi';

export const START_FETCHING_PLANETS = 'START_FETCHING_PLANETS';
export const DONE_FETCHING_PLANETS = 'DONE_FETCHING_PLANETS';
export const SET_PLANETS = 'SET_PLANETS';
export const FAILED_FETCHING_PLANETS = 'FAILED_FETCHING_PLANETS';

export type SetPlanetsParams = {
    count: number;
    nextPage: number;
    results: Planet[];
}

type PlanetActionTypes = typeof SET_PLANETS
    | typeof DONE_FETCHING_PLANETS
    | typeof START_FETCHING_PLANETS
    | typeof FAILED_FETCHING_PLANETS;

export type FetchPlanetsAction = {
    type: PlanetActionTypes;
}
