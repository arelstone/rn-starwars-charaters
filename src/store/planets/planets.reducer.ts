import { Planet } from '../../SwApi';
import { START_FETCHING_PLANETS, DONE_FETCHING_PLANETS, SET_PLANETS, FAILED_FETCHING_PLANETS } from './planet.types';
import { StoreResource } from '../root.reducer';


export const INITIAL_STATE: StoreResource<Planet[]> = {
    count: 0,
    loading: false,
    nextPage: undefined,
    results: [],
    error: {
        hasError: false,
        message: '',
    },
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Planet[]> => {
    switch (type) {
        case START_FETCHING_PLANETS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PLANETS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PLANETS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

