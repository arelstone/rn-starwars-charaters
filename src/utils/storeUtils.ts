export const getNumber = (string: string) => Number(string.replace(/\D/g, ''));
