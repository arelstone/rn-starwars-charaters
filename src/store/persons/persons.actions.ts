import { Person, ResourceType } from '../../SwApi';
import { getNumber, nextPage } from '../../utils/storeUtils';
import { FAILED_FETCHING_PERSONS, START_FETCHING_PERSONS, SET_PERSONS, DONE_FETCHING_PERSONS } from './persons.types';
import { baseUrl } from '../../config';

export const fetchPersons = (page = 1, uri: ResourceType = ResourceType.People) => async (dispatch: any) => {
    if (page === 1) {
        dispatch({ type: START_FETCHING_PERSONS });
    }

    const response = await fetch(`${baseUrl()}/${uri}/?page=${page}`)
        .then(r => r.json())
        .catch(error => dispatch({
            type: FAILED_FETCHING_PERSONS,
            payload: error.message,
        }));

    const next = nextPage(response);

    dispatch({
        type: SET_PERSONS,
        payload: {
            nextPage: next,
            count: response.count,
            results: response.results.map((person: Person) => ({
                ...person,
                id: getNumber(person.url),
                homeworld_id: getNumber(person.homeworld),
            }))
                .sort((a: Person, b: Person) => a.name.toLowerCase().localeCompare(b.name.toLowerCase())),
        },
    });

    if (next) {
        return dispatch(fetchPersons(next));
    }
    console.log('###HERE', next);


    return dispatch({ type: DONE_FETCHING_PERSONS });
};
