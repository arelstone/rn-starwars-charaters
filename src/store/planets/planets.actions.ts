import { Dispatch } from 'react';
import { GetState } from '..';
import { Planet } from '../../SwApi';
import { getNumber } from '../../utils/storeUtils';
import { SetPlanetsParams, FetchPlanetsAction } from './planet.types';

export const fetchPlanets = (page = 1) => async (dispatch: Dispatch<FetchPlanetsAction>, getState: GetState) => {
    if (page === 1) {
        dispatch({ type: 'START_FETCHING_PLANETS' });
    }
    const response = await fetch(`https://swapi.dev/api/planets/?page=${page}`)
        .then(r => r.json())
        .catch(error => console.log('###error', error));

    const nextPage = response.next && getNumber(response.next);

    const results = response.results.map((result: Planet) => ({
        ...result,
        id: getNumber(result.url),
    }));

    dispatch(setPlanets({
        nextPage: response.next && getNumber(response.next),
        results: [...results, ...getState().planets.results],
        count: response.count,
    }));

    // if (nextPage) {
    //     return dispatch(fetchPlanets(nextPage));
    // }

    return dispatch({ type: 'DONE_FETCHING_PLANETS' });
};

const setPlanets = (payload: SetPlanetsParams) => dispatch => {
    return dispatch({
        type: 'SET_PLANETS',
        payload,
    });
};
