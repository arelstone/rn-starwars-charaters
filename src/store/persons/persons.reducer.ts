import { Person } from '../../SwApi';
import { StoreResource } from '../root.reducer';
import { DONE_FETCHING_PERSONS, FAILED_FETCHING_PERSONS, SET_PERSONS, START_FETCHING_PERSONS } from './persons.types';

export const INITIAL_STATE: StoreResource<Person[]> = {
    count: 0,
    loading: false,
    error: {
        hasError: false,
        message: '',
    },
    nextPage: undefined,
    results: [],
};

export default (state = INITIAL_STATE, { type, payload }: any): StoreResource<Person[]> => {
    switch (type) {
        case START_FETCHING_PERSONS:
            return {
                ...state,
                results: [],
                loading: true,
                error: {
                    hasError: false,
                    message: '',
                },
            };
        case DONE_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
            };
        case FAILED_FETCHING_PERSONS:
            return {
                ...state,
                loading: false,
                error: {
                    hasError: true,
                    message: payload,
                },
            };
        case SET_PERSONS: {
            return {
                ...state,
                ...payload,
                results: [
                    ...state.results,
                    ...payload.results,
                ],
            };
        }
        default:
            return state;
    }
};

