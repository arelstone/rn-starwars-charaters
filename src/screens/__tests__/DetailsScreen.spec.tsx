import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { DetailsScreen } from '../DetailsScreen';
import { navigationMock } from '../../../__mocks__/navigation.mock';
import { personsMock } from '../../../__mocks__/persons.mock';
import { planetsMock } from '../../../__mocks__/planets.mock';
import { findElementByTestId } from '../../utils/testUtils';

let wrapper: ShallowWrapper;
const charater = personsMock.find(p => p.id === 1);
const planet = planetsMock.find(p => p.id === 1);

describe('<DetailsScreen />', () => {
    beforeAll(() => {
        wrapper = shallow(<DetailsScreen
            navigation={navigationMock}
            route={{
                key: '',
                name: 'Details',
                params: {
                    id: 1,
                    planetId: 1,
                },
            }}
            {...{ charater, planet }}
        />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should display the name of the Charater', () => {
        expect(findElementByTestId('Text', 'CharaterName', wrapper).at(0).prop('children')).toEqual(charater?.name);
    });

    it('should display the name of the Planet', () => {
        expect(findElementByTestId('Text', 'PlanetName', wrapper).at(0).prop('children').includes(planet?.name)).toEqual(true);
    });

    it('should display the name of the HairColor', () => {
        expect(findElementByTestId('Text', 'HairColor', wrapper).at(0).prop('children').includes(charater?.hair_color)).toEqual(true);
    });

    it('should display the name of the EyeColor', () => {
        expect(findElementByTestId('Text', 'EyeColor', wrapper).at(0).prop('children').includes(charater?.eye_color)).toEqual(true);
    });

    it('should display the name of the SkinColor', () => {
        expect(findElementByTestId('Text', 'SkinColor', wrapper).at(0).prop('children').includes(charater?.skin_color)).toEqual(true);
    });


});
