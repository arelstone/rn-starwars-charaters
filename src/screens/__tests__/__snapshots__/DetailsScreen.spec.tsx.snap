// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsScreen /> should match snapshot 1`] = `
<Screen
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 450,
        "position": "absolute",
        "right": 450,
        "top": 450,
      }
    }
  >
    <Avatar
      gender="male"
      hairColor="Blond"
      skinColor="Fair"
    />
  </View>
  <View>
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
      testID="Charatername"
    >
      Luke Skywalker
    </Text>
    <Text
      testID="PlanetName"
    >
      Lives on: 
      Tatooine
    </Text>
    <Text
      testID="Gender"
    >
      Gender: 
      male
    </Text>
    <Text
      testID="HairColor"
    >
      Hair color: 
      Blond
    </Text>
    <Text
      testID="EyeColor"
    >
      Eye color: 
      Blue
    </Text>
    <Text
      testID="SkinColor"
    >
      Skin color: 
      Fair
    </Text>
  </View>
</Screen>
`;
